### PR TITLE
Revise RFC 0800 with learnings from implementation

### PR DESCRIPTION
**Pull request by chriskrycho**
This revises the original RFC #800 with several points learned during implementation:

- Instead of implementing this RFC *in terms of* `ember-cli-typescript`, we plan to sunset it. We will write a corresponding blog post and coordinate with the CLI team about timing to finish up a transition that really began several years ago.

- We learned that keeping around the service registry is useful for integration with `Owner`, and implemented that in terms that make it user extensible (this needs further documentation, but is well-defined).

- We answered the unresolved questions: how to start projects, updates for blueprints, how we handle classic features, how we integrate “constructibility” as defined in #730, and how we plan (at a high level) to do side-by-side TS and JS docs.

### CODE REVIEW

File: text/0800-ts-adoption-plan.md
Comment (dfreeman) [inline text/0800-ts-adoption-plan.md:570]: I _think_ the current status quo is that TS transpilation is enabled out-of-the-box with Embroider, but for classic `ember-cli-babel` you still have to opt in via `enableTypeScriptTransform: true` config.

CLI folks, what would you think of changing that default to `true` in the v8 beta?

**Review by wagenet — COMMENTED** [review]
Seems generally good to me.

File: text/0800-ts-adoption-plan.md
Comment (NullVoxPopuli) [inline text/0800-ts-adoption-plan.md:577]: > can ship blueprints which support both TypeScript and JavaScript

is this easy? or documented?

Or is it just the concept of:
 - author your blueprints in TS
 - manually transpile out the TS via babel?

Comment (chriskrycho) [reply text/0800-ts-adoption-plan.md:577]: It’s built into Ember CLI (shipped last year! It was part of #776), and it’s easy. You author in TS and the consumer gets TS or JS and it Just Works™ for consumers. We just (!) need to document it in the CLI docs.

Comment (chriskrycho) [reply text/0800-ts-adoption-plan.md:577]: And to be clear: that “just works” for *all* addons since it landed; it’s not specific to `ember-source` or anything.

File: text/0800-ts-adoption-plan.md
Comment (bertdeblock) [reply text/0800-ts-adoption-plan.md:570]: > I think the current status quo is that TS transpilation is enabled out-of-the-box with Embroider

Is this true?
I was setting up TS in an Embroider project, and without the flag, I'm getting errors like:

```ts
$TMPDIR/embroider/9437f6/routes/application.ts/application.ts: Unexpected token, expected "from" (4:12)

  2 | import Route from '@ember/routing/route';
  3 | import { service } from '@ember/service';
> 4 | import type IntlService from 'ember-intl/services/intl';
    |             ^
  5 | import config from 'ember-project-boilerplate/config/environment';
  6 |
  7 | export default class ApplicationRoute extends Route {
```

It could very well be that I'm misconfiguring something though!

Comment (bertdeblock) [reply text/0800-ts-adoption-plan.md:570]: > CLI folks, what would you think of changing that default to true in the v8 beta?

Seems okay to me, though I don't know if this would have any downsides tbh.

Comment (dfreeman) [reply text/0800-ts-adoption-plan.md:570]: @bertdeblock Ahhh, I had tested that `.ts` files were processed, but I didn't actually include any TypeScript-specific syntax in them. As of https://github.com/embroider-build/embroider/pull/1236 it seems like Embroider unconditionally includes `.ts` source files, but still only actually treats it as TypeScript if `'ember-cli-babel': { includeTypeScriptTransform: true }` is set in the app or v1 addon's options.

> I don't know if this would have any downsides for non-TS projects tbh

Outside of some odd edge cases (like if a project has `.ts` and `.js` files with the same name for some reason, or if an addon is erroneously shipping `.ts` files in its `app/` directory), I think this change should be largely invisible for non-TS projects.

The other scenario this might impact would be if someone is manually configuring the TS transform in their Babel config themselves—in that case, they'd need to explicitly opt out with `enableTypeScriptTransform: false` to continue doing that. If this lands as part of the v8 changeover, though, that seems ok to me.

Comment (ef4) [reply text/0800-ts-adoption-plan.md:570]: If we think there's no downside to enabling it for everyone, then there shouldn't be an option at all. It should just be hard-coded to enabled.

On the other hand if there are legitimate reasons to not want it in an app that doesn't use typescript, and if the default blueprint doesn't use typescript, then I think the default should stay off.

It's true that embroider has some baked-in typescript support, but that is an artifact of our compatibility layers. Where we're heading, your babel config would be completely transparent and it will either have the typescript preset in it or not -- there's no policy question about ember-cli-babel flags.

Comment (bertdeblock) [reply text/0800-ts-adoption-plan.md:570]: Did a quick test with enabling `enableTypeScriptTransform` in a non-TS project.
TS code in `.js` files is also transpiled away and doesn't error.

Comment (NullVoxPopuli) [reply text/0800-ts-adoption-plan.md:570]: I think we should have default on, because it makes choosing TS after a JS project easier

Comment (bertdeblock) [reply text/0800-ts-adoption-plan.md:570]: Having it on by default would be easier indeed. However, it does feel a bit strange that in a non-TS app (where there's no TS related config whatsoever) TS code in a JS file would now also be transpiled away, instead of throwing an error.

Comment (dfreeman) [reply text/0800-ts-adoption-plan.md:570]: > On the other hand if there are legitimate reasons to not want it in an app that doesn't use typescript, and if the default blueprint doesn't use typescript, then I think the default should stay off.

As I said above, my concern isn't apps that don't use TS, but rather apps that _are_ using TS but with other nonstandard Babel config that necessitates manual configuration.

That said, now that we've clarified that Embroider only supports the `.ts` extension and not the syntax by default, I'm not as concerned with changing `ember-cli-babel`'s default.

File: text/0800-ts-adoption-plan.md
Comment (simonihmig) [inline text/0800-ts-adoption-plan.md:568]: > for addons, emitting type declarations during build

Any idea or prescriptions how do that exactly? With e-c-ts we have dedicated commands for this, which I believe put the `.d.ts` files at the root level and know how to properly clean this up again as part of `postpack`. Without it, will we just put a long `tsc` command into `prepack`, mostly replicating [what the command does](https://github.com/typed-ember/ember-cli-typescript/blob/master/ts/lib/commands/precompile.ts#L28-L36)? But then how do we clean this up? Or should we put all declarations into a separate folder (we could use `dist` if it wasn't used already for the Ember CLI build, so maybe `declarations`?) and have `exports` and `typesVersions` configs in `package.json` point to it, similar to what v2 addons already do?

Comment (simonihmig) [inline text/0800-ts-adoption-plan.md:568]: > Configure Glint

I am all up for this _personally_, however I wonder if that is really ready yet for everyone? Like this feels like a substantial change in DX for new addons, so have all the implications of doing this been considered? Maybe this justifies its own RFC even?

For example when you create a new component using `ember g`, it won't type-check with Glint, as the generated test file (which is not using `<template>` yet) would use the new component in the template, which is unknown to Glint as the component blueprint does not (currently - will it ever?) add it to the template registry!

Again, I am not opposed to it _personally_ , just want to call out if this needs further consideration!?

**Review by simonihmig — COMMENTED** [review]
I have this [draft PR](https://github.com/ember-cli/ember-cli/pull/10283) lined up to implement the changes proposed here. While working on this and seeing the result in action two questions came up, which I am leaving here for further discussion...

File: text/0800-ts-adoption-plan.md
Comment (chriskrycho) [reply text/0800-ts-adoption-plan.md:568]: 1. I appreciate that you just solved this while I was buried in other things. 😁

2. I think the approach you landed on over at ember-cli/ember-cli#10283 is a good one. For other folks who haven't read that (and for posterity): the blueprint [emits types into `declarations`][decl] and [uses `typesVersions` to make resolution work correctly for consumers][tv]. (It also removes that after publishing.) This is probably what we would have done in the first place with `ember-cli-typescript` if we had had `typesVersions` available all those years ago.

[decl]: https://github.com/ember-cli/ember-cli/pull/10283/files#diff-30de8a6a8a92275125cfd88e3d43feaed99481211446fe701ce5ec00ff4642edR4
[tv]: https://github.com/ember-cli/ember-cli/pull/10283/files#diff-0ee18a1bec40c2bce48535d75189607df32ab18631183de0297b3c3755d352aaR92-R99

**Comment by gitKrystan** [comment]
Per TS core meeting: @chriskrycho to check status

**Comment by chriskrycho** [comment]
Thanks to folks for (a) feedback and (b) implementing this! I think it's ready to merge, pending a review from @wagenet and/or @gitKrystan.

**Review by gitKrystan — APPROVED** [review]

**Review by wagenet — APPROVED** [review]

**Pull request merged**
